src/human_encoding/parse/ast.rs

```diff
@@ -0,0 +1,846 @@
+// Simplicity "Human-Readable" Language
+// Written in 2023 by
+//   Andrew Poelstra <simplicity@wpsoftware.net>
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+//! Parsing
+
+use std::mem;
+use std::str::FromStr;
+use std::sync::Arc;
+
+use crate::human_encoding::{Error, ErrorSet, Position};
+use crate::jet::Jet;
+use crate::{node, types};
+use crate::{BitIter, Cmr, FailEntropy};
+use santiago::grammar::{Associativity, Grammar};
+use santiago::lexer::{Lexeme, LexerRules};
+
+/// A single non-empty line of a program, of the form x = y :: t
+///
+/// A program is simply a list of such lines
+#[derive(Debug, PartialEq, Eq, Clone, Hash)]
+pub struct Line<J> {
+    /// Position of the first character of the line.
+    pub position: Position,
+    /// The name of the expression being named on the line.
+    pub name: Arc<str>,
+    /// The actual expression, if present (missing for type declarations).
+    pub expression: Option<Expression<J>>,
+    /// The type of the expression, if given (inferred if missing).
+    pub arrow: (Option<Type>, Option<Type>),
+}
+
+/// An expression, as represented in the AST
+#[derive(Debug, PartialEq, Eq, Clone, Hash)]
+pub struct Expression<J> {
+    pub inner: ExprInner<J>,
+    pub position: Position,
+}
+
+impl<J: Jet> Expression<J> {
+    fn reference(name: Arc<str>, position: Position) -> Self {
+        Expression {
+            inner: ExprInner::Reference(name),
+            position,
+        }
+    }
+}
+
+/// An expression, as represented in the AST
+#[derive(Debug, PartialEq, Eq, Clone, Hash)]
+pub enum ExprInner<J> {
+    /// A reference to another expression
+    Reference(Arc<str>),
+    /// A left assertion (referring to the CMR of an expression on the right)
+    AssertL(Arc<Expression<J>>, AstCmr<J>),
+    /// A right assertion (referring to the CMR of an expression on the left)
+    AssertR(AstCmr<J>, Arc<Expression<J>>),
+    /// An inline expression
+    Inline(node::Inner<Arc<Expression<J>>, J, node::NoDisconnect, node::NoWitness>),
+}
+
+/// A CMR, as represented in the AST
+#[derive(Debug, PartialEq, Eq, Clone, Hash)]
+pub enum AstCmr<J> {
+    Expr(Arc<Expression<J>>),
+    Literal(Cmr),
+}
+
+/// A type, as represented in the AST
+#[derive(Debug, PartialEq, Eq, Clone, Hash)]
+pub enum Type {
+    /// A named type variable
+    Name(String),
+    /// The unit type 1
+    One,
+    /// The bit type 1+1
+    Two,
+    /// A product type (A * B)
+    Product(Box<Type>, Box<Type>),
+    /// A sum type (A + B)
+    Sum(Box<Type>, Box<Type>),
+    /// An exponential type 2^n where n is one of 2, 4, 8, 16, ..., 1024.
+    TwoN(u32),
```

Comment:
8e711fd6b8a34c454d7ec83a8ad672cc6a8e12f1: `TwoN(n)` means 2^n many bits. `Two(9)` is converted to the Simplicity type `2^512` and so on. We could rename it to `Bits` and update the documentation to make this clear. I was really confused why `TwoN(n)` was passed as `two_two_n(n)` and everything worked.